### DISCUSSION TITLE
How to configure the application sync period from the default of 3 minutes?

### DISCUSSION THREAD

**Discussion by toast-gear** [argoproj/argo-cd#8538] [Q&A]
The reconciliation logic will perform a re-sync on each application every 3 minutes, how can I configure this to be lower (or higher). Additionally, what is the lowest sync period possible

**Comment by toast-gear** [comment]
I believe I've found how:
https://github.com/argoproj/argo-cd/blob/master/docs/operator-manual/upgrading/2.0-2.1.md#replacing---app-resync-flag-with-timeoutreconciliation-setting -> https://github.com/argoproj/argo-cd/blob/master/docs/operator-manual/argocd-cm.yaml#L262

**Reply by mnarrell** [reply]
Does this work?  No matter what I do to this field, the reconciliation remains 3m.

**Reply by ptchau2003** [reply]
Same here :(

**Reply by alxivanov** [reply]
Hi, if you installed Argo using operator, try this:
```
$ oc edit argocd 
...
apiVersion: argoproj.io/v1alpha1
kind: ArgoCD
metadata:
  ...
  name: argocd
  ...
spec:
  controller:
    env:
    - name: ARGOCD_RECONCILIATION_TIMEOUT
      value: 60s
...
```
After you save your changes, you can verify:
```
$ oc get po
NAME                                  READY   STATUS      RESTARTS   AGE
argocd-application-controller-0       1/1     Running     0          17s

$ oc set env sts/argocd-application-controller --list
# statefulsets/argocd-application-controller, container argocd-application-controller
ARGOCD_RECONCILIATION_TIMEOUT=60s

$ oc describe app test-app | grep "Reconciled At:"
  Reconciled At:          2023-01-05T09:43:27Z
$ oc describe app test-app | grep "Reconciled At:"
  Reconciled At:          2023-01-05T09:44:27Z
```

Use the value of `0` to disable reconciliation by timeout. Hope this helps.

**Reply by Ognian** [reply]
Tried to disable with `0s` but it simply doesn't work. Tried syncWindows too, no change either.

**Reply by alxivanov** [reply]
@Ognian, try with the value of `0` instead of `0s`. According to the documentation, it is this value that disables the functionality. Not sure it will work, though, as I have not tried it myself.

**Reply by Ognian** [reply]
@alxivanov thanks a lot for your quick answer, but unfortunately it doesn't help either.
with `0` the following warning is thrown:
```
time="2023-01-08T16:49:15Z" level=warning msg="Could not parse '0' as a duration string from environment ARGOCD_RECONCILIATION_TIMEOUT"                                                                                                                                      │
│ time="2023-01-08T16:49:15Z" level=info msg="ArgoCD Repository Server is starting" built="2022-12-16T16:10:27Z" commit=fc3eaec6f498ddbe49a5fa9d215a219191fba02f port=8081 version=v2.5.5+fc3eaec                  
```
with `0s` there is no such warning
but in both cases I can still see the access of the codecommit repo in the cloud trail event log...

and additionally I have a deny service window in my default project
<img width="1020" alt="image" src="https://user-images.githubusercontent.com/1231174/211211911-89d08b4e-3f07-43f2-9a43-42c0ea0876e6.png">
which is of no help either.

**Reply by alxivanov** [reply]
Strange. I think you should report that as a bug to the maintenance team.

**Reply by jannfis** [reply]
The reconciliation is not supposed to be disabled. If you set it to `0s`, it will use the default value (which is `3m`). You could set it to a very high value (e.g. `24h`),  but not disable it.

**Reply by Ognian** [reply]
@jannfis is there any other way to "disable" argocd and if needed easily 'enable' it? The reconciliation causes real costs (it count as a repo user)... I'm having projects where development and maintenance are "frozen", so this reconciliation is causing costs without any surplus... So turning off the reconciliation would be a much needed feature.

**Reply by jannfis** [reply]
@Ognian you can always delete the `Application` object (without deleting the application resources) to make Argo CD unaware of something. If you disable the reconciliation completely, for everything, what value would a running Argo CD be in this scenario?

Maybe [this PR](https://github.com/argoproj/argo-cd/pull/11879) will help you? It aims to let you disable reconciliation on a per-application basis instead of globally.

**Reply by Ognian** [reply]
@jannfis I wanted to disable the automatic periodic reconciliation, I was hoping that a manual refresh would still trigger it... But of course one solution may be to scale all argocd pods to 0 and if I would need argocd again I could scale them up again...

**Reply by jannfis** [reply]
Ah OK, I understand. You want to be fully manual (i.e. refresh manually only as well). Sorry, hadn't thought about that.

I think there was some kind of reasoning about not being able to disable the reconciliation completely - @alexmt might have some insights here.

**Reply by alxivanov** [reply]
@jannfis Regarding value of `0` to disable reconciliation by timeout: I have also tried using it without seeing the desired effect (as you said, it defaulted back to 3 minutes), though the docs kind of say that it should:
https://github.com/argoproj/argo-cd/blob/master/docs/operator-manual/argocd-cm.yaml#L284

However, I did manage to disable it with the following:
```
spec:
  controller:
    env:
    - name: ARGOCD_RECONCILIATION_TIMEOUT
      value: 0s
```
so I think I was not reading the documentation correctly. By the way, the value of `timeout.reconciliation` in the argocd-cm `configmap` resource is still ignored.

@Ognian, after applying the above configuration I was able to issue a sync command to an app that was out of sync to get it syncronized and it worked as expected:
```
$ argocd app sync test-app --server=$(oc get route argocd-server -o jsonpath='{.spec.host}') | grep Message:
Message:            successfully synced (no more tasks)
$ oc describe app | grep "Reconciled At:"
  Reconciled At:          2023-01-16T09:42:05Z
  Reconciled At:          2023-01-16T09:42:03Z
  Reconciled At:          2023-01-16T10:57:27Z
  Reconciled At:          2023-01-16T09:42:03Z
  Reconciled At:          2023-01-16T09:42:03Z
  Reconciled At:          2023-01-16T09:42:03Z
```
(note the time difference of the 3rd one from the top). 
**EDIT:** "A manual refresh" is still possible for me through the web UI (simply selected the app and clicked "Refresh" button), or through Argo CD CLI:
```
$ argocd app get test-app --refresh --server=$(oc get route argocd-server -o jsonpath='{.spec.host}') | grep "Sync Status:"
Sync Status:        OutOfSync from HEAD (12bbe21)
```
To manually refresh all apps in a namespace, you can do something similar to this:
```
$ for app in $(oc get app --no-headers | awk '{print $1}'); do argocd app get $app --refresh --server=$(oc get route argocd-server -o jsonpath='{.spec.host}') | grep "Sync Status:"; done
Sync Status:        Synced to HEAD (d822ae1)
Sync Status:        OutOfSync from HEAD (12bbe21)
Sync Status:        Synced to HEAD (12bbe21)
Sync Status:        Synced to HEAD (a0f031f)
Sync Status:        Synced to HEAD (a0f031f)
Sync Status:        Synced to HEAD (a0f031f)
```

**Reply by Ognian** [reply]
@alxivanov indeed setting the env var of the controller like you wrote above disables the reconciliation and the git repo is not accessed anymore (I verified also via aws cloud trail ). But I was not able to issue a manual refresh. In the WebUI when clicking on refresh in an app the refresh button gets disabled and the refresh animation starts, but it never ends... Looks like this setting prevents the entire access to the git repo, regardless of automatic or manual...
@jannfis what do you think? is there any chance to get the desired behavior?

**Reply by alxivanov** [reply]
@Ognian Sorry for getting back to you this late, but I have now checked and it turns out that it is possible to issue a manual refresh in the web UI (clicking the refresh button) when reconciliation is disabled. It is therefore hard for me to say what the issue is on your end. Have you tried the CLI method I described earlier?

Here is how it looks in my project now:
```
$ oc describe app | grep "Reconciled At:"
  Reconciled At:          2023-01-16T12:08:25Z
  Reconciled At:          2023-01-16T12:13:21Z
  Reconciled At:          2023-01-16T12:08:27Z
  Reconciled At:          2023-01-20T14:16:12Z
  Reconciled At:          2023-01-20T14:16:13Z
  Reconciled At:          2023-01-25T08:34:48Z
```

**Reply by Ognian** [reply]
I found now little time to dig a little bit deeper:
I'm using now the latest helm chart and the latest argo-cd version.
The actual Problem is that I'm not able setting `ARGOCD_RECONCILIATION_TIMEOUT` to any value.
The reason is that if I try to set via:
```
spec:
  controller:
    env:
    - name: ARGOCD_RECONCILIATION_TIMEOUT
      value: 0s
```
this leads to to an error when starting up the statefullset **argo-cd-argocd-application-controller** the error is:
```
statefulset-controller  create Pod argo-cd-argocd-application-controller-0 in StatefulSet argo-cd-argocd-application-controller failed error: Pod "argo-cd-argocd-application-controller-0" is invalid: spec.containers[ │
│ [].env[0].valueFrom: Invalid value: "": may not be specified when `value` is not empty 
```
 and if we look [here](https://github.com/argoproj/argo-helm/blob/0845b604805ba32db3994e0759d46108dc18428b/charts/argo-cd/templates/argocd-application-controller/statefulset.yaml#L100) we know that due to the fact that this value should come from the config map it is impossible to be overwritten this way.

So the only way of setting this is via the config map. Now the config map looks like this:
<img width="548" alt="image" src="https://user-images.githubusercontent.com/1231174/224777513-834261e4-d67e-4659-af30-87a477d6a4f9.png">

but the statefullset still has this env var not set correctly:
<img width="1289" alt="image" src="https://user-images.githubusercontent.com/1231174/224778017-e890297d-d9cb-439f-9049-a0b483294081.png">

I have no idea what is causing this...

**Reply by Ognian** [reply]
OK looks like I have to kill the pod of the statefullset even I've rebooted my 1 node cluster :-) ...
After this the env var is indeed set correctly and then the output of the argo-cd-argocd-application-controller is:
<img width="1465" alt="Pasted Graphic 8" src="https://user-images.githubusercontent.com/1231174/224801344-fd00a5b9-deef-4f9c-a551-48583dba4b70.png">
So the reconcile timeout is set to 876000 hours which should be around 100 years which propably could be enough time ...
Manual refresh looks to work, I'm checking with
` kubectl describe app -A | grep "Reconciled At:"`
I have to check the logs of my repo to see if this indeed works, but at the moment it looks good.
